File: fjalekryq-client/src/app/core/services/coin.service.ts
import { Injectable, signal } from '@angular/core';

const COINS_KEY      = 'fjalekryq_coins';
const LAST_LOGIN_KEY = 'fjalekryq_last_login';
const STREAK_KEY     = 'fjalekryq_login_streak';

const STARTING_COINS = 100;
const DAILY_REWARDS  = [20, 25, 30, 40, 50, 75, 100]; // day 1–7 of streak

export const HINT_COST  = 10;
export const SOLVE_COST = 50;

@Injectable({ providedIn: 'root' })
export class CoinService {
  readonly coins = signal(0);

  constructor() {
    const stored = localStorage.getItem(COINS_KEY);
    if (stored === null) {
      // First ever launch — give starting bonus
      this.coins.set(STARTING_COINS);
      this.save();
    } else {
      this.coins.set(Math.max(0, parseInt(stored, 10) || 0));
    }
  }

  canAfford(amount: number): boolean {
    return this.coins() >= amount;
  }

  add(amount: number): void {
    this.coins.update(v => v + amount);
    this.save();
  }

  spend(amount: number): boolean {
    if (!this.canAfford(amount)) return false;
    this.coins.update(v => v - amount);
    this.save();
    return true;
  }

  /**
   * Call once per session (e.g. level-map open).
   * Returns reward info if today hasn't been claimed yet, null otherwise.
   * Coins are added automatically when a reward is returned.
   */
  claimDaily(): { amount: number; day: number } | null {
    const today = new Date().toDateString();
    const lastLogin = localStorage.getItem(LAST_LOGIN_KEY);
    if (lastLogin === today) return null; // already claimed today

    const yesterday = new Date();
    yesterday.setDate(yesterday.getDate() - 1);
    const streak = parseInt(localStorage.getItem(STREAK_KEY) ?? '0', 10);
    const newStreak = lastLogin === yesterday.toDateString() ? (streak % 7) + 1 : 1;

    try {
      localStorage.setItem(LAST_LOGIN_KEY, today);
      localStorage.setItem(STREAK_KEY, String(newStreak));
    } catch { /* ignore */ }

    const amount = DAILY_REWARDS[newStreak - 1];
    this.add(amount);
    return { amount, day: newStreak };
  }

  private save(): void {
    try { localStorage.setItem(COINS_KEY, String(this.coins())); } catch { /* ignore */ }
  }
}

File: fjalekryq_mobile/lib/core/services/coin_service.dart
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';

const String _coinsKey = 'fjalekryq_coins';
const String _lastLoginKey = 'fjalekryq_last_login';
const String _streakKey = 'fjalekryq_login_streak';

const int startingCoins = 100;
const int hintCost = 10;
const int solveCost = 50;

const List<int> dailyRewards = [20, 30, 45, 60, 80, 100, 125];

/// Manages the coin balance, spending, earning, and daily rewards.
/// Ported from coin.service.ts
class CoinService extends ChangeNotifier {
  final SharedPreferences _prefs;
  int _coins = 0;

  CoinService(this._prefs) {
    final stored = _prefs.getInt(_coinsKey);
    if (stored == null) {
      _coins = startingCoins;
      _save();
    } else {
      _coins = stored < 0 ? 0 : stored;
    }
  }

  int get coins => _coins;

  bool canAfford(int amount) => _coins >= amount;

  void add(int amount) {
    _coins += amount;
    _save();
    notifyListeners();
  }

  bool spend(int amount) {
    if (!canAfford(amount)) return false;
    _coins -= amount;
    _save();
    notifyListeners();
    return true;
  }

  /// Check if today's reward is available without claiming it.
  ({int amount, int day})? peekDaily() {
    final today = _todayString();
    if (_prefs.getString(_lastLoginKey) == today) return null;

    final yesterday = _yesterdayString();
    final lastLogin = _prefs.getString(_lastLoginKey);
    final streak = _prefs.getInt(_streakKey) ?? 0;
    final newStreak = (lastLogin == yesterday) ? (streak % 7) + 1 : 1;
    return (amount: dailyRewards[newStreak - 1], day: newStreak);
  }

  /// Current claimed streak day (1-7).
  int get currentStreakDay {
    final s = _prefs.getInt(_streakKey) ?? 1;
    return s < 1 ? 1 : s;
  }

  /// Claim today's reward. Returns reward info or null if already claimed.
  ({int amount, int day})? claimDaily() {
    final today = _todayString();
    final lastLogin = _prefs.getString(_lastLoginKey);
    if (lastLogin == today) return null;

    final yesterday = _yesterdayString();
    final streak = _prefs.getInt(_streakKey) ?? 0;
    final newStreak = (lastLogin == yesterday) ? (streak % 7) + 1 : 1;

    _prefs.setString(_lastLoginKey, today);
    _prefs.setInt(_streakKey, newStreak);

    final amount = dailyRewards[newStreak - 1];
    add(amount);
    return (amount: amount, day: newStreak);
  }

  void _save() {
    _prefs.setInt(_coinsKey, _coins);
  }

  String _todayString() {
    final now = DateTime.now();
    return '${now.year}-${now.month}-${now.day}';
  }

  String _yesterdayString() {
    final yesterday = DateTime.now().subtract(const Duration(days: 1));
    return '${yesterday.year}-${yesterday.month}-${yesterday.day}';
  }
}
